docs/trie.md

```diff
@@ -0,0 +1,77 @@
+# Trie
+
+We use Merkle-Patricia Trie to store blockchain state.
+Trie is persistent, which means that insertion of new node actually leads to creation of new path to this node, and thus root of Trie after insertion will also be presented by new object.
+
+Here we describe its implementation details which are closely related to Runtime.
+
+## Main structures
+
+### Trie
+
+Base structure. Almost never called directly from `Runtime`, modifications are made using `TrieUpdate`.
+
+### TrieUpdate
+
+Provides a way to access Storage and record changes to commit in the future. Update is prepared as follows:
+
+- changes are made using `set` and `remove` methods, which are added to `prospective` field,
+- call `commit` method which moves `prospective` changes to `committed`,
+- call `finalize` method which prepares `TrieChanges` and state changes based on `committed` field.
+
+Note that `finalize` does not modify `Trie` itself, it only prepares changes to be passed to `ChainStore`.
+`Trie::insert` and `Trie::update` also does not modify `Trie`. The former one puts new nodes to temporary memory object, and the latter one prepares `TrieChanges` object.
+
+`Trie` is actually updated when `ShardTries::apply_insertions` is called, which puts new values to `ColState` part of key-value database.
```

Comment:
Here, the docs would also benefit from explaining *why* before *how*. So it makes sense to start with something along the lines of "we want to make changes to the trie with the transactional semantics. That is, we want to be able to start transaction, accumulate a bunch of changes, and either apply them all or rollback them all. To achieve that, we buffer all the changes in memory, then we compute the actual diff we need to apply to the data on disk, and finally we write the data out" (note that this is how I'd imagine this to work, might be completely off base). While we are at it, it will also be helpful to say *when* we start transaction. Are we running a trie transaction per block? Or are we running a trie transaction per block transaction? Can we have more than one transaction open?

Reply:
In fact, I also don't currently know all the reasons of current implementation.
My objective is to explain some details I was closely working with and which were not easily understandable from the code, though I certainly will spend some time on improving explanations.